### PR TITLE
Add unicode clipboard support for Desktop Access

### LINKED ISSUE #12906: Clipboard support for unicode text
Support unicode text for copy/paste.

### LINKED ISSUE #12909: Clipboard text encoding mismatch
When cutting/copying from browser-client to remote-Windows-desktop,  Clipboard text is [encoded as UTF-8 on the client](https://github.com/gravitational/webapps/blob/04c7af4065a26e2eb79fff8821daf4348d6e178b/packages/teleport/src/lib/tdp/codec.ts#L345), and then passed through RDP to the Windows desktop by [adding CR/LF line endings](https://github.com/gravitational/teleport/blob/master/lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs#L106), and then [sending it as `CF_OEMTEXT`](https://github.com/gravitational/teleport/blob/master/lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs#L137).

However it seems as though UTF-8 is not a subset of [`CF_OEMTEXT`](https://docs.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats). For example, a `'` copied on the client will appear as `ΓÇÖ` on the Windows desktop (and vice versa if you go the other direction).

### PR DESCRIPTION
**Pull request by xacrimon**
This PR commits some sins to make pasting unicode from the browser work correctly, the whole flow can be a bit tricky to follow so here's what happens for a typical paste from a browser now:
- Browser clipboard contains data in UTF-16
- Our client encodes it as UTF-8 and sends it over the wire to the proxy
- The proxy picks it up and calls `Client::update_clipboard`
- The client checks if the UTF-8 is pure ASCII (they're backwards compatible):

If yes:
  - Set format to `CF_TEXT`
  - Add 1 byte nul terminator

If no:
  - Reencode text as UTF-16LE
  - Set format in clipboard to `CF_UNICODETEXT`
  - Add 2 byte nul terminator

On clipboard formats and Windows code pages:
After doing some research, it seems that `CF_OEMTEXT` uses a mostly archaic Windows OEM codepage that is seldom used nowadays, switching to `CF_TEXT` will use the extended ANSI codepage instead which has better support and works with characters like '. This should still be compatible with apps that rely on a non unicode clipboard format. If we fetch a `CF_OEMTEXT` clipboard from the server anyway, we attempt to just interpret it as UTF-8 which is in most cases valid.

Fixes #12906
Fixes #12909

### CODE REVIEW

**Comment by zmb3** [comment]
I am not opposed to having a separate TDP message for "Unicode Clipboard Data" and making the current one only for ASCII text if that simplifies things. What do you think?

**Comment by xacrimon** [comment]
@zmb3 I don't really agree. First of all, we'd still be transmitting text in the same format, UTF-8 from client to proxy since UTF-8 is backwards compatible with ascii, secondly, we'd need to perform this analysis anyway to validate the contents of the message before we ship it off to the RDP server. Due to this, the gains from a seperate TDP message kinds seem marginal and I don't think it'd actually make anything easier to the server, perhaps trading an if statement or two for a new message handler.

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 1]
```diff
@@ -103,38 +103,32 @@ impl Client {
...
-        const CR: u8 = 13;
-        const LF: u8 = 10;
-
-        // convert LF to CRLF, as required by CF_OEMTEXT
+        // convert LF to CRLF, as required by CF_TEXT and CF_UNICODETEXT
         let len_orig = data.len();
         let mut converted = Vec::with_capacity(len_orig);
         for i in 0..len_orig {
             match data[i] {
-                LF => {
```
Comment (xacrimon) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:115]: I felt this looked clearer but open to feedback.

Comment (zmb3) [reply lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:115]: No objections from me.

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 2]
```diff
@@ -548,6 +545,22 @@ impl<T: FormatName> FormatListPDU<T> {
     }
 }
 
+fn encode_clipboard(mut data: Vec<u8>) -> RdpResult<(Vec<u8>, ClipboardFormat)> {
+    if data.is_ascii() {
+        if matches!(data.last(), Some(x) if *x != 0) {
+            data.push(0);
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:551]: Can we add a comment just so it's clear that this is adding a 1-byte null terminator?

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 3]
```diff
@@ -1051,38 +1064,40 @@ mod tests {
...
-            ("\n123", "\r\n123\0"),   // starts with LF
-            ("def\r\n", "def\r\n\0"), // already CRLF, no conversion necessary
-            ("gh\r\nij\nk", "gh\r\nij\r\nk\0"), // mixture of both
+        struct Item(&'static [u8], &'static [u8], ClipboardFormat);
+        for Item(input, expected, format) in [
+            Item(b"abc\0", b"abc\0", ClipboardFormat::CF_TEXT), // already null-terminated, no conversion necessary
+            Item(b"\n123", b"\r\n123\0", ClipboardFormat::CF_TEXT), // starts with LF
+            Item(b"def\r\n", b"def\r\n\0", ClipboardFormat::CF_TEXT), // already CRLF, no conversion necessary
+            Item(b"gh\r\nij\nk", b"gh\r\nij\r\nk\0", ClipboardFormat::CF_TEXT), // mixture of both
+            Item(
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:1087]: Would be good to add a test that is UTF-8 (not ASCII) and also needs CRLF conversion.

**Comment by zmb3** [comment]
What happens if the browser sends invalid UTF-8 data to us? Do we handle this gracefully or does the session terminate?

**Comment by xacrimon** [comment]
@zmb3 The session terminates, we could just ignore it and drop the message which might be a better option, or alternatively attempt a lossy conversion (I like this best now that I think about it)

**Comment by espadolini** [comment]
What happens with the official RDP client and server if you try to paste some broken UTF8?

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 4]
```diff
@@ -103,38 +103,32 @@ impl Client {
     /// update_clipboard is invoked from Go.
     /// It updates the local clipboard cache and returns the encoded message
     /// that should be sent to the RDP server.
+    ///
+    /// The data passed to this function should be valid UTF-8 text.
     pub fn update_clipboard(&mut self, data: Vec<u8>) -> RdpResult<Vec<Vec<u8>>> {
```
Comment (espadolini) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:108]: If broken utf 8 will result in an error, why not just take a `String` here instead?

File: lib/srv/desktop/rdp/rdpclient/src/util.rs [hunk 5]
```diff
@@ -22,7 +22,6 @@
 /// convert it to RDP's preferred format and send it over the wire.
 pub fn to_unicode(s: &str) -> Vec<u8> {
     let mut buf: Vec<u8> = s.encode_utf16().flat_map(|v| v.to_le_bytes()).collect();
```
Comment (espadolini) [inline lib/srv/desktop/rdp/rdpclient/src/util.rs:24]:
```suggestion
    s.encode_utf16().chain([0]).flat_map(|v| v.to_le_bytes()).collect()
```

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs (hunk 3 above)
Comment (xacrimon) [reply lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:1086]: Addressed

File: lib/srv/desktop/rdp/rdpclient/src/util.rs [hunk 6]
```diff
@@ -21,8 +21,8 @@
 /// to handle some data in the code as a &str (or String), and later
 /// convert it to RDP's preferred format and send it over the wire.
 pub fn to_unicode(s: &str) -> Vec<u8> {
```
Comment (espadolini) [inline lib/srv/desktop/rdp/rdpclient/src/util.rs:23]: nit: `to_nul_terminated_utf16le`?

**Review by espadolini — APPROVED** [review]

**Comment by zmb3** [comment]
image

I was able to paste Wejdenstål, so it must work, right? 😄

**Comment by xacrimon** [comment]
@zmb3 Ah, tricky. I'll look into fixing that.

**Comment by xacrimon** [comment]
I've now added code pick the optimal format when receiving a paste and proper decoding. Skies should be blue now but I have yet to get AD working again to test this.

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 7]
```diff
@@ -271,21 +272,19 @@ impl Client {
...
+        let resp = FormatDataResponsePDU::decode(payload, length)?;
         let data_len = resp.data.len();
+        let format = self.incoming_paste_formats.pop_front().ok_or_else(|| {
+            try_error("no expected format found, possibly received too many format data responses")
+        })?;
+
         debug!(
-            "recieved {} bytes of copied data from Windows Desktop",
-            data_len,
+            "recieved {} bytes of copied data from Windows Desktop with format {:?}",
```
Comment (espadolini) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:282]:
```suggestion
            "received {} bytes of copied data from Windows Desktop with format {:?}",
```

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 8]
```diff
@@ -548,6 +547,54 @@ impl<T: FormatName> FormatListPDU<T> {
...
+// This data must be valid UTF-8.
+fn encode_clipboard(mut data: Vec<u8>) -> (Vec<u8>, ClipboardFormat) {
+    if data.is_ascii() {
+        if matches!(data.last(), Some(x) if *x != b'\0') {
+            data.push(b'\0');
+        }
+
+        (data, ClipboardFormat::CF_TEXT)
+    } else {
+        let str = std::str::from_utf8(&data).unwrap();
```
Comment (espadolini) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:562]: I have to say, not a big fan of this safely accessible panic here 😬

Are you sure there's no clean way to do the LF->CRLF translation while still staying in the comfy realm of `String`/`str`?

**Review by espadolini — APPROVED** [review]

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs (hunk 8 above)
Comment (xacrimon) [reply lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs]: I've now made this conversion without going to bytes and got rid of the unwrap.

**Review by zmb3 — APPROVED** [review]
Just gave this another try and it works in both directions now. Thanks!

**Pull request merged**